src/libstd/sys/windows/fs.rs

```diff
@@ -567,23 +646,85 @@ pub fn remove_dir_all(path: &Path) -> io::Result<()> {
         // rmdir only deletes dir symlinks and junctions, not file symlinks.
         rmdir(path)
     } else {
-        remove_dir_all_recursive(path)
+        // canonicalize to get a `//?/`-path, which can handle files like `CON`
+        // and `morse .. .`,  and when a directory structure is so deep it needs
+        // long path names
+        let path = try!(path.canonicalize());
+        let base_dir = match path.parent() {
+            Some(dir) => dir,
+            None => return Err(io::Error::new(io::ErrorKind::PermissionDenied,
+                                              "can't delete root directory"))
+        };
+        try!(remove_dir_all_recursive(path.as_ref(), base_dir.as_ref(), 0));
+        Ok(())
     }
 }
 
-fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
+fn remove_dir_all_recursive(path: &Path, base_dir: &Path, mut counter: u64)
+        -> io::Result<u64> {
+    // On Windows it is not enough to just recursively remove the contents of a
+    // directory and then the directory itself. Deleting does not happen
+    // instantaneously, but is scheduled. So just before or after flagging a
+    // file for deletion, we move it to some `base_dir` to avoid races.
+
+    fn move_item(file: &File, base_dir: &Path, mut counter: u64)
+            -> io::Result<u64> {
+        let mut tmpname = base_dir.join(format!{"rm-{}", counter});
+        counter += 1;
+        // Try to rename the file. If it already exists, just retry with an other filename.
+        while let Err(err) = file.rename(tmpname.as_ref(), false) {
+            if err.kind() != io::ErrorKind::AlreadyExists { return Err(err) };
+            tmpname = base_dir.join(format!{"rm-{}", counter});
+            counter += 1;
+        }
+        Ok(counter)
+    }
+
+    fn remove_item(path: &Path,
+                   base_dir: &Path,
+                   mut counter: u64,
+                   readonly: bool) -> io::Result<u64> {
+        if !readonly {
+            let mut opts = OpenOptions::new();
+            opts.access_mode(c::DELETE);
+            opts.custom_flags(c::FILE_FLAG_BACKUP_SEMANTICS | // delete directory
+                              c::FILE_FLAG_OPEN_REPARSE_POINT | // delete symlink
+                              c::FILE_FLAG_DELETE_ON_CLOSE);
```

Comment:
Just to confirm, the mode here is:

* This will fail if there are any other option handles which do *not* specific `FILE_SHARE_DELETE`
* All other future openings will fail unless they specify `FILE_SHARE_DELETE` if this succeeds.
* Once all handles are closed, the file is deleted

So in other words, so long as nothing deadlocks, we're guaranteed to delete the file if the open succeeds, right?